### DISCUSSION TITLE
Use cases for custom services?

### DISCUSSION THREAD

**Discussion by gustojs** [feathersjs/feathers#2435] [Q&A] [answered]
Hi, I've been wondering lately, what are some common or example use cases for writing your own custom services or customizing them by extending the service class, rather than providing the functionality via service hooks?
The official guide provides an example when we want to add a link to Gravatar, but that seems like something we still can do in hooks. 
In what situations do you go beyond hooks and customize the service class itself in your own applications?
Were there some cases when you've first opted to write a custom service / extend a service, but later deemed it unnecessary?

(inspired by the current Slack discussion https://feathersjs.slack.com/archives/C0HJE6A65/p1629626955210200)

**Comment by daffl** [comment, marked as answer]
Personally I don't think I ever found a custom service unnecessary later but I've seen more than a few example where what is done in hooks should really be a custom or customized (extended) service instead. This is part of the reason why I added the [Hooks vs. extending services](https://docs.feathersjs.com/guides/basics/hooks.html#hooks-vs-extending-services) section to the guide.

For example, if you end up setting `context.result` in hooks a lot (other than e.g. a generic caching system) it might be worth customizing the service instead. The most extreme example I've seen was a custom service method simply returning `true` which never got hit due to a hook around it always setting `context.result`. That hook should have really been the service method itself.

I have a feeling that the upcoming [schema definitions and resolvers](https://github.com/feathersjs/feathers/issues/2312) will add a more concise way to encapsulate things that used to be done across (multiple) hooks. My hope is it may become more clear where things should go even though it is introducing a new concept:

- Core logic in services (standard db adapter, customized or custom)
- General purpose workflows in hooks (permissions, logging, profiling, sending emails)
- Data modelling and validation in schemas
- Context dependent (and possibly asynchronous) data processing in resolvers (e.g. associations/`$populate`, per property-permissions, password hashing, query restrictions etc.)

**Comment by gustojs** [comment]
Thanks. I've got my hopes high for schema definitions and (especially) resolvers coming up as a new standard for things like population. Good luck with that!

**Comment by bwgjoseph** [comment]
I've used custom service as a service that is not tied to any model (database tables). For example, I want to expose a API that calls other service

```js
// in client
feathers.service('batch').create({ service: 'post', data: [...] });

// in batch service
create(data) {
  // you can do some other stuff here
  this.app.service(data.service).create(data.data);
  // or here before
  return
}
```

So essentially, you exposed one service to your consumer. This is very simple example, but hope you get the gist of it.

**Reply by gustojs** [reply]
Thanks. So one of the use cases is to create wrapper services that abstract some functionality, potentially so that the client doesn't have to be aware of it.

**Reply by bwgjoseph** [reply]
Yup, sort of.  Could be a aggregate service that calls few other service internally rather than client calling multiple service.

**Comment by daffl** [comment]
Another big one I can think of is using a custom service as a proxy to another existing API which allows you to do things like:

- Automatically add real-time events 
- Authentication and authorization proxy
- Smart caching to take the load off the existing API
- Automatically implement REST API best practises
- Automatically add new functionality like Swagger or GraphQL

**Reply by bwgjoseph** [reply]
That's right! I almost forgot about this. I had a service that proxy to other API while providing WS back to my client
